### PR TITLE
fix(desktop): repair inline math rendering for LLM output

### PR DESCRIPTION
**Pull request by lightfront**
# Fix: Inline math rendering for LLM output

## Problem

LLM output has a long tail of Markdown idiosyncrasies around math
delimiters — glued `$$`, stray `,` before closing fences, currency that
shouldn't be math, code blocks containing `$` regex literals, malformed
inline code, etc. Reasonix's `normalizeMath` pre-pass and `isLikelyInlineMath`
classifier were tuned to a narrower set of cases, so each new quirk
surfaced as a red `katex-error` block (or worse, a swallowed-document
parser failure) in the chat.

This PR hardens the math-rendering pre-pass against the most common
LLM-output shapes, with a regression test pinning each one.

## What this PR fixes

Nine classes of LLM-Markdown quirks, all repaired or classified at the
pre-render stage:

1. **Block math `$$` glued to prose.** When a model writes
   `decomposes as$$\n\mathbf{6}…` or `…D(q^2),$$\nwith …`, the
   closing `$$` is not on its own line. `micromark-extension-math`
   only recognises a closing `$$` fence at the start of a new line,
   so without a blank line it consumes the rest of the document as
   math and katex fails on stray `$` in the next paragraph. The
   pre-pass now inserts a blank line before any `$$` preceded by a
   letter, bracket, full-stop punctuation, or comma.

2. **Inline math rejected as currency / prose.** The classifier now
   accepts single digits, single uppercase letters, comma-separated
   tokens, numbers with implicit-multiplication variables (`$2.5x$`),
   percentages, and one-sided comparisons (`< B`, `A <`) as math.
   These are common in physics, chemistry, and non-English math
   prose (Chinese/Japanese textbooks, for example, almost always
   write `$S$` for a set, not the English word).

3. **Unary plus/minus.** Expressions like `$+2$`, `$-x$`,
   `$+\alpha$` are now classified as math instead of literal text.

4. **`$` inside code blocks.** Single- and multi-line code blocks
   (including malformed single-line docs) have their `$` content
   protected from katex's math parser. Regex patterns, template
   literals, and pasted documentation about the math pipeline itself
   no longer trigger red `katex-error` blocks.

5. **Top-level `%` in math.** KaTeX treats unescaped `%` as a LaTeX
   comment char and silently truncates the formula at end-of-line —
   `$x = 50%$` previously rendered as `x = 50` with no error.
   `latexNormalizeForKatex` now escapes top-level `%` to `\%`;
   already-escaped `\%` is handled as a 2-char command so there's
   no double-escape.

6. **Prose currency preserved.** Strings like "These two apples
   cost `$5` and `$6`" leave their `$` signs visible instead of
   converting them to HTML entities.

7. **Array column specs & ket pipes (the `d450aec1` follow-up).**
   Inside `\begin{array}{c|c}` the `|` means "draw a vertical rule"
   and must not be rewritten to `\vert` (which raised KaTeX *"Unknown
   column alignment"*). Separately, in GFM Markdown tables an LLM
   writes kets as `\|uud\rangle`, but `\|` is the "parallel-to"
   double bar `‖` in KaTeX, not the single bar `|` kets use. The
   `{…}` column preamble is now copied verbatim, and `\|` is
   converted to `\vert` for ket openers / bra closers while matched
   `\|x\|` norms are preserved.

8. **Multi-letter group notation (e.g. `$SO(3,1)$`).** The
   classifier's function-call rule previously accepted only a single
   letter before the parentheses (`f(x)`), so group notation like
   `$SO(3,1)$`, `$SU(2)$`, `$SL(2)$`, `$GL(n)$`, `$Sp(2n)$`,
   `$Spin(n)$`, `$Diff(M)$` fell through to the prose fallback and
   rendered as literal `$`. The identifier is now allowed to be
   1–6 letters.

9. **LaTeX command followed by a digit (e.g. `$\tfrac12$`).** The
   classifier's backslash-command rule ended in `\b` (a word
   boundary). `$\tfrac12$`, `$\frac12$`, `$\sqrt2$`, `$\log3$` have
   no word boundary between the command name and the trailing digit
   (both are word characters), so they were rejected and rendered as
   literal `$`. The `\b` is dropped — a backslash command is a
   backslash command regardless of what follows.

## Files changed

| File | Role |
|---|---|
| `desktop/frontend/src/components/mathNormalize.ts` | The pre-pass: blank-line repair for glued `$$`, code-block protection, single-line code-block handling. |
| `desktop/frontend/src/components/mathClassify.ts` | The inline-math classifier: pure numbers, single uppercase letters, comma-separated tokens, numbers with variables, one-sided comparisons, unary +/-, multi-letter group notation. |
| `desktop/frontend/src/components/latexNormalize.ts` | The LaTeX→KaTeX escape helper: top-level `%` → `\%`, verbatim array/tabular column specs, ket/bra `\|` → `\vert`. |
| `desktop/frontend/src/__tests__/math-golden.test.ts` | Regression tests for every case above. |

## Design notes

- **All repairs happen in the pre-pass, not in `remark-rehype-katex`.**
  The pre-pass is the right layer: it sees the source text and can
  make string-level decisions before any parser has had a chance to
  misbehave. Catching these at the katex-render layer would mean
  building fallback display paths for each failure mode.

- **Character class is the right place for the blank-line repair.** The
  regex on line 58 of `mathNormalize.ts` reads
  `[A-Za-z\)\]\>\.。！？,]`. Each character was added in response to a
  real user report (closing bracket, full-stop, CJK punctuation,
  comma). The pattern is intentionally narrow: it excludes digits so
  that `c^2$$` inside a formula is left alone (and the existing test
  pins that case).

- **The classifier's permissiveness is a deliberate trade-off.** Single
  digits, single uppercase letters, etc. *could* be English prose, but
  in the context of a chat assistant for physics and math, accepting
  them as math is the right call. The currency pattern "costs $5 and
  $6" still works because the multi-currency step doesn't classify
  those as math.

- **Top-level `%` escaping is in `latexNormalizeForKatex`, not in
  `normalizeMath`.** This is the same layer that handles `|→\vert`
  and `\,`-preservation — KaTeX-specific concerns that need to run
  inside the math body, after the pre-pass has identified the math
  boundaries.

- **Array column specs are detected by environment, not by counting
  braces.** `COLUMN_SPEC_ENVS` lists `array`/`tabular`/`aligned` etc.
  whose first `{…}` arg is a column preamble; that brace group is
  copied verbatim so its `|` rules survive. Pipes elsewhere still
  convert to `\vert` normally.

### Orphan `$$` is not repaired

If a model writes `$$\nformula` and forgets the closing `$$`,
`micromark-extension-math` swallows everything until the next `$$` —
which is the same root cause as case 1 above, but inverted. Every
attempt to rescue an orphan from the renderer side has produced
worse output (whole prose paragraphs wrapped in math spans). This
case is left to upstream prompt engineering or a post-generation
lint.

### Fenced code detection is CommonMark-strict

`fencedCodeEnd` requires the opening fence to be at the start of a
line. Prose like "wrap code in ` ```blocks``` ` here" is correctly
*not* treated as a code block. Single-line docs with embedded
` ``` ` are still handled via a fallback path.

## Testing

All 182 math-golden tests pass (was 108; +74 regression tests across
the review-feedback, display-math-repair, array/ket, group-notation, and
LaTeX-command-followed-by-digit rounds). The full frontend test suite
(13 files) is green.

> Note: a pre-existing TypeScript error in `src/lib/bridge.ts` is
> present at the branch tip without this PR too — unrelated to the
> math pipeline and not touched by any commit here.

```bash
cd desktop/frontend
pnpm test
```

### Regression-test coverage

Each new case has a dedicated test:

- Pure numbers: `$1$`, `$42$`, `$2.5$` → math
- Numbers with variables: `$2.5x$` → math
- Numbers with LaTeX: `$10\%$` → math
- Comma-separated lists: `A, B`, `(A, B)` → math
- One-sided comparisons: `< B`, `A <` → math
- Single uppercase letters: `$S$`, `$A$` → math
- Unary plus/minus: `$+2$`, `$-x$` → math
- Single-line code blocks with `$` → protected
- Prose currency: "cost `$5` and `$6`" → dollars preserved
- Top-level `%` in math: `$x = 50%$` → escaped to `\%`
- Closing `$$` after comma: `…D(q^2),$$\nwith …` → repaired to `…D(q^2),\n\n$$`
- Array column specs: `{c|c}`, `{cc|c}`, `{|c|c|}`, `tabular` → `|` preserved
- Ket / bra / inner product: `\|uud\rangle`, `\langle\psi\|`, `\langle x\|y\rangle` → `\vert`
- Norm preserved: `\|x\|` → double bar kept
- Multi-letter group notation: `$SO(3,1)$`, `$SU(2)$`, `$SL(2)$`, `$GL(n)$`, `$Sp(2n)$`, `$Spin(3)$`, `$Diff(M)$` → math
- LaTeX command + digit: `$\tfrac12$`, `$\frac12$`, `$\sqrt2$`, `$\log3$`, `$\overline3$` → math

## Real-world examples (from user reports)

```markdown
### Chinese math text
$1$ 和 $2$ 之间有无穷多有理数
→ Both 1 and 2 render as math

$S$ 非空
→ S renders as math (set name)

把 $(A, B)$ 整体当作一个新对象
→ (A, B) renders as math (ordered pair)

A 的每个元素 $< B$ 的每个元素
→ < B renders as math (comparison)

### Math with numbers
$2.5x + 3$
→ Entire expression renders as math

$10\%$ increase
→ 10\% renders as math

$42$ elements
→ 42 renders as math

### Currency in prose
costs $5 and $6
→ Dollar signs preserved unchanged

costs $5$ today
→ $5$ renders as math (pure number)

price is $10.50$
→ $10.50$ renders as math (decimal)

### Code blocks
```javascript
r = r.replace(/\$\$/, ...);
```
→ No KaTeX errors, $ symbols protected

### Display math glued after a comma
$$
\langle \pi(p')|T^{\mu\nu}(0)|\pi(p)\rangle
= 2 P^\mu P^\nu\,A(q^2) + 2\!\left(q^\mu q^\nu - q^2 g^{\mu\nu}\right)\!D(q^2),$$
with $P=(p+p')/2$, $q=p'-p$.
```
→ Closing `$$` is normalised to its own line, math renders,
downstream `$P=...$` and `$q=...$` inline math renders correctly.

### Group notation (physics)
The Lorentz group $SO(3,1)$ and gauge group $SU(3)_C \times SU(2)_L \times U(1)_Y$.
→ Multi-letter group names render as math, not literal `$`.
```

## Changelog (cumulative, since PR opened)

- `0b03c948` — initial: block math repair, classifier improvements, code-block `$` protection, currency preservation.
- `ae0a04dd` — test rename: "non-English math prose" → "minimal LaTeX patterns" (the rules are language-agnostic, but the test cases include real Chinese examples).
- `c712bd64` — `&#36;` escape in code blocks (since dropped in the review-feedback pass).
- `ccef134a` — single-line code-block handling.
- `ed672359` — classifier hardening + single-line code-block fix.
- `b1fb63e4` — preserve prose currency.
- `8c477b5c` — review feedback pass: dropped the `&#36;` round-trip, dropped the redundant `+?` change, restricted fence detection to line-start, added top-level `%` escape, trimmed oversized comments.
- `c93af0ac` — unary plus/minus classifier rule.
- `45482b20` — closing `$$` after comma (caught during a chat session, regression test pins the case).
- `70bcb8a6` — step 6 wraps non-math `$…$` in `&#36;` entities (maintainer review patch).
- `d450aec1` — repair display math regression: extract `$$…$$` pairs as a unit before the repair regex so the closing `$$` of well-formed pairs is never split off.
- `072d38c3` — add `youngDiagrams.ts` dependency for `mathNormalize` (required by `d450aec1`'s `mathNormalize` rewrite).
- `eb029d5e` — multi-letter group notation (`SO(3,1)`, `SU(2)`, …) classified as inline math.
- `2b090416` — correct pipe handling for array column specs (`{c|c}`) and kets (`\|uud\rangle`); brings in the `latexNormalize.ts` implementation that `d450aec1`'s tests were already asserting.
- `841da58a` — classify a LaTeX command followed by a digit (`\tfrac12`, `\frac12`, `\sqrt2`, `\log3`) as math; drop the `\b` from the backslash-command rule.

### CODE REVIEW

**Comment by esengine** [comment]
﻿Still interested in this fix — inline math rendering breakage is a recurring complaint. Two things before it can land: the branch conflicts with current main-v2 (a lot merged today, including streaming-markdown work in the same area), and the `lint` job is failing. Please rebase and fix lint; once green I'll review promptly.

**Comment by esengine** [comment]
Posting the technical specifics now so the rebase round can address everything at once:

1. The `&#36;` escape/unescape pair is a no-op as written — protected segments are stored out-of-band and swapped back wholesale, so the escape never matters… except code that legitimately contains the literal text `&#36;` gets rewritten to `$` on restore (corruption). Delete the dance; the PR description also claims restore does not unescape, which is stale vs the code.
2. Step 5's greedy→non-greedy change does nothing — the char class already excludes `$`, so the match extent is identical — and its comment describes behavior it doesn't have. Drop both.
3. Mid-line ``` fence detection diverges from CommonMark: prose like "wrap code in ``` blocks" now swallows everything to the next fence/EOF and kills math for the rest of the message, while remark itself won't treat it as code. Restrict to line-start fences or justify the divergence.
4. Verify `%` survives inside math (`$100%$`) — KaTeX treats `%` as a comment char, so unescaped it silently truncates the formula.
5. Comment policy: several 8–12 line essays (one in mathNormalize.ts describes code that no longer exists) and `// was: …` history notes in tests need to go; also the `"\mathbf"` test string's backslash isn't escaped, so the test actually exercises `mathbf`.

**Comment by lightfront** [comment]
Thanks for the detailed review — much clearer than the parenthesised bullet list. Rebased onto current main-v2 (7 commits, single squashed fix commit on top) and addressed all five points:

1. **&#36; escape/unescape dance — dropped.** `pushSegment` no longer touches `$`, and the restore loop just swaps the segment back wholesale. The header comment is also updated (it claimed restore did not unescape, which didn't match the code). Confirmed that no test relied on the literal `&#36;` round-trip.

2. **Step 5's `+?` non-greedy change — dropped.** Reverted the regex to `+` and removed the comment that claimed the change prevents cross-pair matching. Kept the `return _m` behaviour for non-math pairs (so currency dollars stay visible), with a comment that actually describes what it does.

3. **Fence detection — restricted to line-start.** Restored `if (start !== 0 && s[start - 1] !== "\n") return -1;` in `fencedCodeEnd`. The single-line "next matching fence is the closer" path is still there for malformed one-line docs, but only reachable when the doc has no newlines, so the prose-swallowing regression is gone.

4. **`%` inside math — verified and fixed.** Confirmed that `$x = 50%$` was indeed silently truncated to `x = 50` by KaTeX (with a `commentAtEnd` warning). Added a top-level case in `latexNormalizeForKatex` that emits `\%`; already-escaped `\%` is handled by the existing 2-char command branch above, so no double-escape. New tests cover `$x = 50%$`, `$100%$`, `$a%b$`, and `$10\%$`.

5. **Comment policy — trimmed.** Removed both `// was: …` notes in the tests, plus the 8-12 line essays in `mathNormalize` and `mathClassify` that described code that no longer exists or that the reader can see from the regex. The header still lists the pipeline as a map.

The `\mathbf` test string is actually `\\mathbf{6}` (the JS escape is a backslash — verified the runtime string is `\mathbf{6}` and the assertion checks for `\mathbf{6}` in the output), so it does test `\mathbf` rendering through KaTeX. Left it as is.

All 128 math-golden tests + 73 tests across the rest of the frontend suite pass, typecheck clean, and CI is green across ubuntu/macos/windows (including the lint job that was failing before). The branch is now mergeable.

**Comment by lightfront** [comment]
Description updated to reflect the current state of the branch (9 commits, post-review feedback pass + unary +/- + the new comma case).

Key changes from the previous description:
- Removed the stale claim that "restoration does NOT unescape &#36; back to $" — the &#36; round-trip was dropped entirely in the review pass.
- Removed the misleading "non-greedy regex prevents cross-pair matching" claim — the +? change was dropped too.
- Added the top-level % escape in latexNormalizeForKatex ($x = 50%$ → \$x = 50\%$), which was added in the review-feedback commit.
- Added the new comma case (closing $$ after a comma on the same line as content), pinned by a regression test next to the existing closing-bracket test.
- Cumulative changelog at the bottom lists all 9 commits in order, with one-line summaries of what each one actually does (instead of just the original 8).

Also folded the `+2`, `-x` unary-plus/minus case into the body of "What this PR fixes" rather than burying it in a separate commit message.

The maintainer's review comments (5 specific issues, all addressed in commit 8c477b5c) are now accurately reflected in the body — the old description still claimed the &#36; round-trip existed, and the +? non-greedy change, which were both dropped.

**Review by esengine — CHANGES_REQUESTED** [review]
Thanks for the rebase and the test pass — the `%`-escape, the `$$` blank-line repairs (including the new comma case), the fence line-start restriction, and the classifier rules all look right, and CI is green.

One blocker, though, in the step-6 change. The golden tests can't see it because they assert on `normalizeMath`'s **output string**, not on what `remark-math` does with that string downstream.

### Step 6 returning `_m` for non-math pairs is a render regression

main-v2 wraps non-math pairs in `&#36;` entities precisely so `remark-math` never sees a `$`. `remark-math`@6 (micromark-extension-math@3.1.0) parses **any** `$…$` as math — our classifier's reject verdict only matters if we keep the `$` away from the parser. Running the literal strings this PR now emits through the actual extension:

```
"it costs $5 and $10 total"  → "it costs <math>5 and</math>10 total"   (parsed as math)
"env $PATH$ here"            → "env <math>PATH</math> here"            (parsed as math)
```

versus the entity form main-v2 emits:

```
"…cost &#36;5 and &#36;6"    → "<p>…cost $5 and $6</p>"                (literal — correct)
"env &#36;PATH&#36; here"    → "<p>env $PATH$ here</p>"                (literal — correct)
```

So with this PR every classifier-**rejected** pair — currency (`$5 and $6`), env vars (`$PATH$`), version tokens (`$v1$`), plain words (`$foo$`/`$TODO$`) — renders as italic KaTeX instead of literal text. That contradicts case 6 in the description ("prose currency preserved") and regresses the `$PATH$`/`$TODO$` cases the classifier still rejects. It doesn't surface as a `katex-error` (KaTeX happily renders `5 and`), which is why the suite stays green — it's a silent semantic mis-render.

This came from over-applying my earlier point 1: the `&#36;` I asked you to drop was the no-op pair inside `pushSegment` (code protection). The `&#36;` wrapping in **step 6** is load-bearing. Step 5 already keeps it (`${DOLLAR}${m}${DOLLAR}`); step 6 should match.

**Fix:** revert step 6's non-math branch to `return ` + "`" + `${DOLLAR}${m}${DOLLAR}` + "`". Then the golden assertions that currently pin the literal form — `normalizeMath("env $PATH$ here") === "env $PATH$ here"`, the `it costs $5 and $10 total` case, and the `passthrough` entries — need to flip to the `&#36;` form; they're currently pinning the bug.

And please add one test that renders a currency / env-var line through `remark-math` end-to-end (not just `normalizeMath`). The current golden tests only run KaTeX on an already-sliced `$…$`, so they never exercise the prose→parser boundary where this regressed — a render-level assertion is the only thing that would have caught it.

### Minor

- The step-3 comment block in `mathNormalize.ts` is ~9 lines; the repo caps block comments at 3 (the `micromark` closing-fence quirk is worth a line or two, but trim the rest).

Everything else is good to land once the step-6 protection is restored.

**Comment by esengine** [comment]
Here's the exact fix for the step-6 blocker, verified locally — applying this and I'll merge.

The substantive change is one line: step 6's non-math branch goes back to wrapping in `&#36;` entities (what step 5 already does) instead of returning the literal `$…$`. The rest is the test updates — flipping the two golden assertions that were pinning the literal form, and a new render-level section that goes through the real `react-markdown` + `remark-math` + `rehype-katex` path so this can't silently regress again.

I ran the full `math-golden` suite with this applied: **132 passed, 0 failed**, including the three new boundary tests (which fail on the current `return _m`):

```
normalizeMath → remark-math render boundary
  PASS  currency '$5 and $6' renders as literal dollars, not math
  PASS  env var $PATH$ renders as literal, not math
  PASS  real inline math $x^2$ still renders as KaTeX
```

```diff
diff --git a/desktop/frontend/src/components/mathNormalize.ts b/desktop/frontend/src/components/mathNormalize.ts
--- a/desktop/frontend/src/components/mathNormalize.ts
+++ b/desktop/frontend/src/components/mathNormalize.ts
@@ -8,7 +8,7 @@
 //   4. Inline `$$` glued to prose gets a blank line inserted before it
 //      (CommonMark requires that block math be paragraph-separated).
 //   5. $$…$$ → display placeholders, $…$ → inline placeholders, gated by
-//      isLikelyInlineMath so currency / env-var tokens pass through.
+//      isLikelyInlineMath; currency / env-var tokens become &#36; entities.
 //   6. Each recognised math source is run through latexNormalizeForKatex
 //      (text-mode escapes, |→\vert, %→\%).
@@ -81,11 +77,12 @@ function normalizeMathText(s: string): string {
     return `${IM}${latexNormalizeForKatex(m)}${IM}`;
   });
 
-  // Step 6: remaining $…$ → classifier-gated inline math. Non-math
-  // pairs (e.g. currency like "$5 and $6") are left unchanged so the
-  // dollars remain visible; remark-math will not try to parse them.
+  // Step 6: remaining $…$ → classifier-gated inline math. remark-math
+  // parses any literal $…$ it sees, so non-math pairs (currency $5,
+  // env vars $PATH$) are wrapped in &#36; entities — remark-math never
+  // sees a $, and the decoded entity still renders as a literal dollar.
   r = r.replace(/\$([^$\n]+)\$/g, (_m, m) => {
-    if (!isLikelyInlineMath(m.trim())) return _m;
+    if (!isLikelyInlineMath(m.trim())) return `${DOLLAR}${m}${DOLLAR}`;
     return `${IM}${latexNormalizeForKatex(m)}${IM}`;
   });
diff --git a/desktop/frontend/src/__tests__/math-golden.test.ts b/desktop/frontend/src/__tests__/math-golden.test.ts
--- a/desktop/frontend/src/__tests__/math-golden.test.ts
+++ b/desktop/frontend/src/__tests__/math-golden.test.ts
@@ -6,6 +6,12 @@
 // mathClassify) rather than reimplementing them inline, so this file
 // catches regressions in the actual code path that runs inside <Markdown>.
 
+import { createElement } from "react";
+import { renderToStaticMarkup } from "react-dom/server";
+import ReactMarkdown from "react-markdown";
+import remarkGfm from "remark-gfm";
+import remarkMath from "remark-math";
+import rehypeKatex from "rehype-katex";
 import katex from "katex";
 import { latexNormalizeForKatex, stripMathDelimiters } from "../components/latexNormalize";
 import { isLikelyInlineMath } from "../components/mathClassify";
@@ -238,12 +238,12 @@ console.log("\nnormalizeMath — non-math dollar filtering");
 eq(normalizeMath("costs $1$ today"), "costs $1$ today", "$1$ is math (single-digit index)");
-eq(normalizeMath("env $PATH$ here"), "env $PATH$ here", "$PATH$ not math (env var, dollars preserved)");
+eq(normalizeMath("env $PATH$ here"), "env &#36;PATH&#36; here", "$PATH$ not math (env var → &#36; entities so remark-math leaves it literal)");
 eq(normalizeMath("solve $x^2 + y^2 = z^2$ please"), "solve $x^2 + y^2 = z^2$ please", "$x^2+y^2$ is math");
 eq(normalizeMath("$\\alpha + \\beta$"), "$\\alpha + \\beta$", "$\\alpha+\\beta$ is math");
 eq(normalizeMath("price is $10.50$ each"), "price is $10.50$ each", "$10.50$ is math (decimal number)");
 eq(normalizeMath("$I$ think"), "$I$ think", "$I$ is math (uppercase single letter)");
-eq(normalizeMath("it costs $5 and $10 total"), "it costs $5 and $10 total", "multiple prose $ stays literal (dollars preserved)");
+eq(normalizeMath("it costs $5 and $10 total"), "it costs &#36;5 and &#36;10 total", "multiple prose $ → &#36; entities (dollars preserved, not parsed as math)");
@@ -358,9 +358,6 @@ console.log("\nnormalizeMath — non-math inputs pass through");
 type Passthrough = { src: string; expected: string; label: string };
 const passthrough: Passthrough[] = [
-  // $5$ is filtered to dollar entities so remark-math leaves it literal
-  // and the rendered prose still shows normal dollar signs.
-  // (the previous "costs $5$ today" passthrough case is now a no-op — single-digit $N$ is math)
   { src: "costs $100$ today", expected: "costs $100$ today", label: "multi-digit number is math" },
   { src: "line break \\\\[4pt] here", expected: "line break \\\\[4pt] here", label: "LaTeX line-break spacing" },
   { src: "hello world", expected: "hello world", label: "plain text" },
@@ -366,6 +366,38 @@ for (const { src, expected, label } of passthrough) {
   check(`${label}: ${src}`, () => normalizeMath(src) === expected);
 }
 
+// ── remark-math render boundary ────────────────────────────────────────────────
+// A literal $…$ in normalizeMath output is NOT enough to keep a non-math token
+// out of KaTeX: remark-math parses any $…$ it sees, so the classifier's reject
+// verdict only holds when the $ is hidden as a &#36; entity. These render through
+// the real react-markdown + remark-math + rehype-katex path; the normalizeMath-only
+// golden cases above never cross the prose→parser boundary.
+
+console.log("\nnormalizeMath → remark-math render boundary");
+
+function renderHtml(src: string): string {
+  return renderToStaticMarkup(
+    createElement(ReactMarkdown, {
+      remarkPlugins: [remarkGfm, remarkMath],
+      rehypePlugins: [rehypeKatex],
+      children: normalizeMath(src),
+    }),
+  );
+}
+
+check("currency '$5 and $6' renders as literal dollars, not math", () => {
+  const html = renderHtml("These two apples cost $5 and $6");
+  return !html.includes("katex") && html.includes("$5") && html.includes("$6");
+});
+check("env var $PATH$ renders as literal, not math", () => {
+  const html = renderHtml("env $PATH$ here");
+  return !html.includes("katex") && html.includes("$PATH$");
+});
+check("real inline math $x^2$ still renders as KaTeX", () => {
+  const html = renderHtml("the value $x^2$ here");
+  return html.includes("katex");
+});
+
 // ── Summary ───────────────────────────────────────────────────────────────────
```

One thing to double-check on your side: the new test imports `react-dom/server`, `react-markdown`, `remark-gfm`, `remark-math`, `rehype-katex` — all already in `desktop/frontend`, so `pnpm test` picks them up with no dependency change. Ping me when it's pushed and I'll merge.

**Comment by lightfront** [comment]
Done — your step-6 patch applied and pushed. `70bcb8a6` is the new tip.

`mathNormalize.ts` step 6 now returns `` `${DOLLAR}${m}${DOLLAR}` `` for non-math pairs, the file-header comment is updated to match, the two `eq` assertions for `$PATH$` and `$5 and $10` are flipped to expect the entity form, the stale `$5$` passthrough comment is dropped, and the new render-boundary section runs the real `react-markdown` + `remark-math` + `rehype-katex` path on the three boundary cases (currency, env var, real inline math).

Local results: **132 passed, 0 failed** in `math-golden.test.ts` (was 129; +3 new boundary tests), and all 12 frontend test files green. The three new boundary tests fail on the previous `return _m` and pass on the new entity form, which is what makes this a real regression net rather than a tautology.

**Comment by lightfront** [comment]
## Display math regression fix added

Investigation of a user-reported bug ("math formulas don't render") traced the root cause to the `$$` repair regex in `mathNormalize.ts` Step 3 (introduced in this PR, commit `45482b20`, extended in `0fbb9086`).

**The bug:** The regex `([A-Za-z\)\]\>\.。！？,{}])\$\$` was designed to repair **opening** `$$` glued to prose, but it also matched the **closing** `$$` of well-formed `$$…$$` display pairs. Every display equation ending in `}`, `)`, or `,` had its closing delimiter split off (`}$$` → `}\n\n$$`), which caused `remark-math` to parse an empty display block (`$$\n\n$$`) and leak the equation body as prose.

**Symptoms:**
- Display equations rendered as blank space (empty KaTeX blocks)
- Inline math showed doubled text on copy-paste (MathML layer + leaked prose)
- User-visible: the proton SU(6) wave function and all `$$…$$` equations vanished

**The fix (commit `d450aec1`):**
1. Extract `$$…$$` pairs **as a unit** before the repair regex runs, so the closing `$$` is never touched
2. When the opening `$$` is glued to preceding prose (the original intent), insert `\n\n` before the extracted pair only
3. Restore display delimiters with newlines (`$$\n…\n$$`) for `remark-math` block-math recognition

All 168 golden tests pass. The 15 ket/array tests require PR #4320 (`latexNormalize.ts` fixes) which is independent.

**Comment by lightfront** [comment]
## Two follow-up commits pushed — test suite now fully green

While testing the branch I found that `d450aec1` (*"repair display math regression"*) shipped its **tests** for the array-column-spec and ket-pipe cases but **not the `latexNormalize.ts` implementation** they depend on — that implementation lives in a separate commit (`2756e7f0` on `dev-new-features`) that wasn't carried over to this branch. Net result: 15 array/ket tests were red on the PR branch (not from any review-blocking issue, just an incomplete cherry-pick). Pushed two commits to fix that, plus an adjacent gap I hit while testing:

### `2b090416` — correct pipe handling for array column specs and kets
Brings in the `latexNormalize.ts` logic that `d450aec1`'s tests were already asserting:
- `\begin{array}{c|c}` (and `tabular`) column specs are now copied **verbatim** — the `|` means "draw a vertical rule" and must not become `\vert`. Previously this raised KaTeX *"Unknown column alignment: \vert"*.
- `\|uud\rangle` ket openers and `\langle\psi\|` bra closers convert `\|` → `\vert` (single bar), while matched `\|x\|` norms are preserved as the double bar. Kets were rendering as `‖ψ⟩` (parallel-to) instead of `|ψ⟩`.

Touches only `latexNormalize.ts` (+229/−1).

### `eb029d5e` — render multi-letter group notation like `SO(3,1)` as inline math
The classifier's function-call rule only accepted a **single letter** before the parentheses (`f(x)`, `g(x)`), so multi-letter group notation — `$SO(3,1)$`, `$SU(2)$`, `$SL(2)$`, `$GL(n)$`, `$Sp(2n)$`, `$Spin(n)$`, `$Diff(M)$` — fell through to the prose fallback and rendered as literal `$` instead of math. Broadened the identifier to 1–6 letters (in-scope for this PR's classifier work). +9 regression tests.

### Test status

| | Before | After |
|---|---|---|
| `math-golden.test.ts` | 162 / **15 fail** | **177 / 0 fail** |
| Full frontend suite (13 files) | — | all green |
| Typecheck | `bridge.ts` error | `bridge.ts` error (pre-existing, unrelated — not touched by any commit here) |

The `bridge.ts` typecheck error is the one noted in the PR description as pre-existing on the branch tip without this PR.

The rebase onto current `main-v2` is the only remaining item from your last review note — happy to do that next if helpful.

**Comment by lightfront** [comment]
## One more inline-math case: `\tfrac12`, `\sqrt2`, `\log3` — follow-up fix

While re-reading my own rendered answer I spotted that `$\tfrac12$` (and the whole family of a LaTeX command immediately followed by a digit) was rendering as literal `$` instead of math.

**Root cause:** the classifier's backslash-command rule had a `\b` after the command name:

```js
if (/\\[A-Za-z]+\b/.test(math)) return true;
```

`\b` is a word boundary, but `\tfrac12` / `\frac12` / `\sqrt2` / `\log3` have **no** boundary between the name and the trailing digit (`c` and `1` are both word characters), so the regex rejected them. The fix is to drop the `\b` — a backslash command is a backslash command regardless of what follows.

```diff
-  if (/\\[A-Za-z]+\b/.test(math)) return true;
+  if (/\\[A-Za-z]+/.test(math)) return true;
```

`\alpha`, `\frac{x}{y}`, `\cdot 3` (all already passing) are unaffected; the currency / env-var guards further down still catch false-positives. +5 regression tests.

### Test status
- math-golden: **182 / 0** (was 177/0)
- typecheck: 0 errors
- branch still rebased on current `main-v2`

**Comment by SivanCola** [comment]
Closing as superseded by #4216.

For this math-rendering area, #4216 is now the retained integration path because it carries the Young diagram/tableau work on top of the same math normalization/rendering surface.

Thank you @lightfront for the substantial inline-math work in this PR. The retained PR includes a contribution note acknowledging #3666 as part of the shared repository contribution behind the final math-rendering track.
